src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -0,0 +1,314 @@
+                f"builds.\n\n"
+                "To use a custom lockfile, set this option to a file path relative to the "
+                f"build root, then run `./pants generate-lockfiles "
+                f"--resolve={cls.options_scope}`.\n\n"
+                "Lockfile generation currently does not wire up the `[python-repos]` options. "
+                "If lockfile generation fails, you can manually generate a lockfile, such as "
+                "by using pip-compile or `pip freeze`. Set this option to the path to your "
+                "manually generated lockfile. When manually maintaining lockfiles, set "
+                "`[python].invalid_lockfile_behavior = 'ignore'`."
+            ),
+        )
+
+    @property
+    def version(self) -> str:
+        return cast(str, self.options.version)
+
+    @property
+    def artifacts(self) -> tuple[Coordinate, ...]:
+        return tuple(
+            Coordinate.from_coord_str(s.replace("%VERSION%", self.version))
+            for s in self.options.artifacts
+        )
+
+    @property
+    def extra_artifacts(self) -> tuple[Coordinate, ...]:
+        return tuple(Coordinate.from_coord_str(s) for s in self.options.extra_artifacts)
+
+    @property
+    def lockfile(self) -> str:
+        f"""The path to a lockfile or special string '{DEFAULT_TOOL_LOCKFILE}'."""
+        lockfile = cast(str, self.options.lockfile)
+        if lockfile != DEFAULT_TOOL_LOCKFILE:
+            return lockfile
+        pkg, path = self.default_lockfile_resource
+        return f"src/python/{pkg.replace('.', '/')}/{path}"
+
+    def lockfile_content(self) -> bytes:
+        lockfile_path = self.lockfile
+        if lockfile_path == DEFAULT_TOOL_LOCKFILE:
+            return importlib.resources.read_binary(*self.default_lockfile_resource)
+        with open(lockfile_path, "rb") as f:
+            return f.read()
+
+    def resolved_lockfile(self) -> CoursierResolvedLockfile:
+        lockfile_content = self.lockfile_content()
+        lockfile_content_json = json.loads(lockfile_content)
+        return CoursierResolvedLockfile.from_json_dict(lockfile_content_json)
+
+
+@union
+class JvmToolLockfileSentinel:
+    options_scope: ClassVar[str]
+
+
+@dataclass(frozen=True)
+class JvmToolLockfileRequest:
+    artifacts: FrozenOrderedSet[Coordinate]
+    resolve_name: str
+    lockfile_dest: str
+
+    @classmethod
+    def from_tool(cls, tool: JvmToolBase) -> JvmToolLockfileRequest:
+        return cls(
+            artifacts=FrozenOrderedSet((*tool.artifacts, *tool.extra_artifacts)),
+            resolve_name=tool.options_scope,
+            lockfile_dest=tool.lockfile,
+        )
+
+
+@dataclass(frozen=True)
+class JvmToolLockfile:
+    digest: Digest
+    resolve_name: str
+    path: str
+
+
+class GenerateJvmLockfilesSubsystem(GoalSubsystem):
+    name = "jvm-generate-lockfiles"
+    help = "Generate lockfiles for JVM tools third-party dependencies."
```

Comment:
Per #13374, we want `generate-lockfiles` to work on all supported platforms, so this may not just be for third-party dependencies.

Reply:
(if it's possible to remove `coursier-resolve` during this work, I'd support that!)

Reply:
See the open question I just posted. https://github.com/pantsbuild/pants/pull/13777#issuecomment-984936883

Reply:
This PR is an intermediate step toward merging everything. The question is not should we merge everything, but rather how PRs to get there.

Reply:
(moving inline)

> Open question: Should this handle user resolve lockfiles? I wanted to reduce the scope of this change so didn't try to merge coursier-resolve yet into the jvm-generate-lockfiles goal.

No, it shouldn't merge with user lockfiles (i.e. `generate-user-lockfile`). But this change _should_ merge with the Python tool lockfile goal: i.e. `generate-lockfiles`. That would align it with the status quo.

Reply:
As discussed offline, fine with doing some of this merging as a fast follow.

Reply:
Want me to handle merging the two goals? I'm now working on lockfiles as my main project so the context switch isn't bad for me